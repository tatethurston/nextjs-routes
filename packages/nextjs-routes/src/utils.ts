import { existsSync, readdirSync, statSync } from "fs";
import { join } from "path";

// istanbul ignore next: io mocking not worthwhile
function findDir(cwd: string, dir: string): string | undefined {
  const paths = [join(cwd, dir), join(cwd, "src", dir)];
  for (const path of paths) {
    if (existsSync(path)) {
      return path;
    }
  }
}

// istanbul ignore next: io mocking not worthwhile
export function getPagesDirectory(cwd: string): string | undefined {
  return findDir(cwd, "pages");
}

// istanbul ignore next: io mocking not worthwhile
export function getAppDirectory(cwd: string): string | undefined {
  return findDir(cwd, "app");
}

// istanbul ignore next: io mocking not worthwhile
export function findFiles(entry: string): string[] {
  return readdirSync(entry).flatMap((file) => {
    const filepath = join(entry, file);
    if (
      statSync(filepath).isDirectory() &&
      !filepath.includes("node_modules")
    ) {
      return findFiles(filepath);
    }
    return filepath;
  });
}

export function isNotUndefined<T>(value: T | undefined): value is T {
  return value !== undefined;
}
