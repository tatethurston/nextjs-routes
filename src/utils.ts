import { existsSync, readdirSync, statSync } from "fs";
import { join, parse } from "path";

const NEXTJS_NON_ROUTABLE_PREFIX = "_";
const DYNAMIC_SEGMENT_RE = /\[(.*?)\]/g;

export function getPagesDirectory(): string {
  if (existsSync("pages")) {
    return "pages";
  }
  return "src/pages";
}

// istanbul ignore next
export function findFiles(entry: string): string[] {
  return readdirSync(entry).flatMap((file) => {
    const filepath = join(entry, file);
    if (
      statSync(filepath).isDirectory() &&
      !filepath.includes("node_modules")
    ) {
      return findFiles(filepath);
    }
    return filepath;
  });
}

type QueryType = "dynamic" | "catch-all" | "optional-catch-all";

interface Route {
  pathname: string;
  query: Record<string, QueryType>;
}

function convertWindowsPathToUnix(file: string): string {
  return file.replace(/\\/g, "/");
}

export function nextRoutes(files: string[]): Route[] {
  const pagesDirectory = getPagesDirectory();
  const filenames = files
    .map(convertWindowsPathToUnix)
    .map((file) =>
      file.replace(pagesDirectory, "").replace(parse(file).ext, "")
    )
    .filter((file) => !parse(file).name.startsWith(NEXTJS_NON_ROUTABLE_PREFIX));

  return filenames.map((filename) => {
    const segments = filename.match(DYNAMIC_SEGMENT_RE) ?? [];
    const query = segments.reduce<Route["query"]>((acc, cur) => {
      const param = cur
        .replace(/\[/g, "")
        .replace(/\]/g, "")
        .replace("...", "");
      let queryType: QueryType = "dynamic";
      if (cur.startsWith("[[")) {
        queryType = "optional-catch-all";
      } else if (cur.startsWith("[...")) {
        queryType = "catch-all";
      }
      acc[param] = queryType;
      return acc;
    }, {});

    const pathWithoutIndexSuffix = filename.replace(/index$/, "");
    const pathWithoutTrailingSlash =
      pathWithoutIndexSuffix.endsWith("/") && pathWithoutIndexSuffix.length > 2
        ? pathWithoutIndexSuffix.slice(0, -1)
        : pathWithoutIndexSuffix;

    return {
      pathname: pathWithoutTrailingSlash,
      query,
    };
  });
}

function getQueryInterface(
  query: Route["query"]
): [query: string, requiredKeys: number] {
  let res = "{ ";
  let requiredKeys = 0;
  Object.entries(query).forEach(([key, value]) => {
    res += key;
    switch (value) {
      case "dynamic": {
        requiredKeys += 1;
        res += ": string";
        break;
      }
      case "catch-all": {
        requiredKeys += 1;
        res += ": string[]";
        break;
      }
      case "optional-catch-all": {
        res += "?: string[] | undefined";
        break;
      }
      // istanbul ignore next
      default: {
        const _exhaust: never = value;
        return _exhaust;
      }
    }
    res += ";";
  });
  res += " }";
  return [res, requiredKeys];
}

export function generate(routes: Route[]): string {
  return `\
// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.'
// Run \`yarn nextjs-routes\` to regenerate this file.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    | ${routes
      .map((route) => {
        const [query, requiredKeys] = getQueryInterface(route.query);
        if (requiredKeys > 0) {
          return `{ pathname: "${route.pathname}"; query: Query<${query}> }`;
        } else {
          return `{ pathname: "${route.pathname}"; query?: Query | undefined }`;
        }
      })
      .join("\n    | ")}

  type Query<Params = {}> = Params & { [key: string]: string | undefined };
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type RouteOrQuery = Route | { query?: { [key: string]: string | undefined } };

  export interface LinkProps extends Omit<NextLinkProps, "href"> {
    href: RouteOrQuery;
  }

  declare function Link(
    props: PropsWithChildren<LinkProps>
  ): DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Route } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type TransitionOptions = Parameters<Router["push"]>[2];

  type Pathname = Route["pathname"];

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  type RouteOrQuery = Route | { query: { [key: string]: string | undefined } };

  export interface NextRouter<P extends Pathname = Pathname>
    extends Omit<Router, "push" | "replace"> {
    pathname: P;
    route: P;
    query: QueryForPathname[P];
    push(
      url: RouteOrQuery,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: RouteOrQuery,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
  }

  export function useRouter<P extends Pathname>(): NextRouter<P>;
}
`;
}
