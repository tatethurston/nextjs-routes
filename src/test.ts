import { nextRoutes, generate } from "./utils.js";

describe("route generation", () => {
  it("typescript", () => {
    const pages = [
      "/404.ts",
      "/[foo].ts",
      "/[foo]/[bar]/[baz].ts",
      "/[foo]/bar/[baz].ts",
      "/[foo]/bar/[baz]/foo/[bar].ts",
      "/[foo]/baz.ts",
      "/_app.ts",
      "/_debug.ts",
      "/_debug/health-check.ts",
      "/_document.ts",
      "/_error.ts",
      "/_error/index.ts",
      "/api/[[...segments]].ts",
      "/api/[...segments].ts",
      "/api/bar.ts",
      "/foo/[slug].ts",
      "/index.ts",
      "/not-found.ts",
      "/settings/bars/[bar].ts",
      "/settings/bars/[bar]/baz.ts",
      "/settings/foo.ts",
      "/settings/index.ts",
    ];
    expect(generate(nextRoutes(pages))).toMatchSnapshot();
  });
});
