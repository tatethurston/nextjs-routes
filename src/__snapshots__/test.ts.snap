// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route generation typescript 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.'
// Run \`yarn nextjs-routes\` to regenerate this file.
/* eslint-disable */

// prettier-ignore
declare module \\"nextjs-routes\\" {
  export type Route =
    | { pathname: \\"/404\\"; query?: Query | undefined }
    | { pathname: \\"/[foo]\\"; query: Query<{ foo: string; }> }
    | { pathname: \\"/[foo]/[bar]/[baz]\\"; query: Query<{ foo: string;bar: string;baz: string; }> }
    | { pathname: \\"/[foo]/bar/[baz]\\"; query: Query<{ foo: string;baz: string; }> }
    | { pathname: \\"/[foo]/bar/[baz]/foo/[bar]\\"; query: Query<{ foo: string;baz: string;bar: string; }> }
    | { pathname: \\"/[foo]/baz\\"; query: Query<{ foo: string; }> }
    | { pathname: \\"/_debug/health-check\\"; query?: Query | undefined }
    | { pathname: \\"/_error\\"; query?: Query | undefined }
    | { pathname: \\"/api/[[...segments]]\\"; query?: Query | undefined }
    | { pathname: \\"/api/[...segments]\\"; query: Query<{ segments: string[]; }> }
    | { pathname: \\"/api/bar\\"; query?: Query | undefined }
    | { pathname: \\"/foo/[slug]\\"; query: Query<{ slug: string; }> }
    | { pathname: \\"/\\"; query?: Query | undefined }
    | { pathname: \\"/not-found\\"; query?: Query | undefined }
    | { pathname: \\"/settings/bars/[bar]\\"; query: Query<{ bar: string; }> }
    | { pathname: \\"/settings/bars/[bar]/baz\\"; query: Query<{ bar: string; }> }
    | { pathname: \\"/settings/foo\\"; query?: Query | undefined }
    | { pathname: \\"/settings\\"; query?: Query | undefined }

  type Query<Params = {}> = Params & { [key: string]: string | undefined };
}

// prettier-ignore
declare module \\"next/link\\" {
  import type { Route } from \\"nextjs-routes\\";
  import type { LinkProps as NextLinkProps } from \\"next/dist/client/link\\";
  import type { PropsWithChildren, MouseEventHandler } from \\"react\\";
  export * from \\"next/dist/client/link\\";

  type RouteOrQuery = Route | { query?: { [key: string]: string | undefined } };

  export interface LinkProps extends Omit<NextLinkProps, \\"href\\"> {
    href: RouteOrQuery;
  }

  declare function Link(
    props: PropsWithChildren<LinkProps>
  ): DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  export default Link;
}

// prettier-ignore
declare module \\"next/router\\" {
  import type { Route } from \\"nextjs-routes\\";
  import type { NextRouter as Router } from \\"next/dist/client/router\\";
  export * from \\"next/dist/client/router\\";
  export { default } from \\"next/dist/client/router\\";

  type TransitionOptions = Parameters<Router[\\"push\\"]>[2];

  type Pathname = Route[\\"pathname\\"];

  type QueryForPathname = {
    [K in Route as K[\\"pathname\\"]]: Exclude<K[\\"query\\"], undefined>;
  };

  type RouteOrQuery = Route | { query: { [key: string]: string | undefined } };

  export interface NextRouter<P extends Pathname = Pathname>
    extends Omit<Router, \\"push\\" | \\"replace\\"> {
    pathname: P;
    route: P;
    query: QueryForPathname[P];
    push(
      url: RouteOrQuery,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: RouteOrQuery,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
  }

  export function useRouter<P extends Pathname>(): NextRouter<P>;
}
"
`;
